Extensibility Examples (#55)
AdobeDocs/commerce-extensibility: src/pages/app-development/examples.md

---
title: Adobe Commerce Extensibility Examples
description: Learn how to use API Mesh, Adobe I/O Events, and App Builder to extend Adobe Commerce.
---

# Extensibility examples

> There are many ways to extend [Adobe Commerce](https://business.adobe.com/products/magento/extended-capabilities.html) today, such as [APIs](https://developer.adobe.com/commerce/webapi/) and [extensions](https://business.adobe.com/products/magento/marketplace.html). However, these options in isolation only encapsulate a small fraction of what is possible with extensibility. This article speculates on what is possible with new and emergent offerings and more complicated integrations to explore the boundaries of what is possible.

Reviewer: I really like this mention of speculation and possibilities. However, "What is possible" is mentioned twice in the last sentence. Maybe can reduce the occurrences or change it up a bit.

## Understanding the components

There are several new and developing technologies created by Adobe that can further extend the Adobe Commerce platform. Before getting into more complex examples, we will take a look at each of these services individually.

### Adobe Developer App Builder

[App Builder](https://developer.adobe.com/app-builder/docs/overview/) is a complete framework that allows you to build and deploy custom web applications to extend Adobe Commerce and other Adobe solutions while running on Adobe infrastructure. With [React Spectrum](https://react-spectrum.adobe.com/react-spectrum/index.html) these single-page applications (SPAs) are only limited by your React knowledge.

> To learn more about App Builder, we recommend following [this step-by-step tutorial](https://developer.adobe.com/app-builder/docs/getting_started/first_app/) on creating your first app. This will provide you with the foundational knowledge of what App Builder is and how it works.

Reviewer: ```suggestion
To learn more about App Builder, follow [this step-by-step tutorial](https://developer.adobe.com/app-builder/docs/getting_started/first_app/) on creating your first app. This provides you with the foundational knowledge of what App Builder is and how it works.
```

To learn more about App Builder, follow [this step-by-step tutorial](https://developer.adobe.com/app-builder/docs/getting_started/first_app/) on creating your first app. These instructions provide you with the foundational knowledge of what App Builder is and how it works.

### API Mesh for Adobe Developer App Builder

[API Mesh](https://developer.adobe.com/graphql-mesh-gateway/) is an Adobe service that allows you to take multiple disparate sources, such as GraphQL endpoints, REST APIs, JSON Schemas, and SOAP implementations, and combine them into a single queryable GraphQL endpoint. By creating a mesh, you can connect Adobe Commerce to App Builder, Adobe IO Gateway, and any other service that provides an accessible source. You can further customize your experience by [transforming](https://developer.adobe.com/graphql-mesh-gateway/gateway/transforms/) source data.

![API Mesh diagram](../_images/api-mesh-diagram.png)
<!-- [Link to .mmd file](../_images/api-mesh-diagram.mmd) -->

> API Mesh has several features which allow for further extensibility:

Reviewer 2: API mesh adds extensibility at the web API layer (headless/consumer level) - its purpose is to provide the relevant data and behavior to the consumer from different API vendors (sources ) without  the need to modify the actual vendors (in most cases you can't even do that because they're 3rd party)
Also usually the consumer is an experience (like we storefront, webstore, kiosk etc)
that's why it outputs graphql - but also because when building a supergraph - graphql allows the consumer to select the data (graphQL query) it wants to consume
API mesh is fit for integrations also (as opposed to be consumed only for end user experiences) or synchronizing data between systems - AEM does that - it literally pulls all products and mixes them into the content - asynchronously and not real time like an experience would.

> - [Hooks](https://developer.adobe.com/graphql-mesh-gateway/gateway/hooks/) - Allow you to invoke composable local or remote functions, which can be used for authenticating or checking for a header before making a request.

Reviewer 2: we only support Before ALL hooks - so that's something to pay attention to
Hooks again is not an industry standard word for extensibility - like for Adobe Commerce we had Plugins - react has hooks - which is the inspiration for this term
So I think a statement like
'Allow you to invoke composable local or remote functions, which can be used for authenticating or checking for a header before making a request.'
doesn't really explain hooks because it doesn't disambiguate the term
and composable local or remote functions - remote functions are not supported only in the TI version - and here we should only document the multitenant version
Also composable is a marketing term that doesn't resonate a lot with developers which is our target audience here - if you must reference it, do so by explaining what's composable about having a multitude of functions when reacting to an action or API event like beforeALL fetch of data

Author: @cpartica I was under the impression that remote functions were supported. It is covered in our public documentation: https://developer.adobe.com/graphql-mesh-gateway/gateway/hooks/#remote-composer-example. @revanth0212

> - [Custom Resolvers](https://developer.adobe.com/graphql-mesh-gateway/gateway/extending-unified-schema/) - Allow you to upload a JavaScript resolver to your mesh that can modify the schema. In [this example](https://developer.adobe.com/graphql-mesh-gateway/gateway/extending-unified-schema/#programmatic-additionalresolvers), we show how you could use a discounts API to apply discounts to your products.

Reviewer 2: upload a JavaScript resolver ->
upload a resolver, written in Javascript 
by the way this version of JS has it's framework and limitations 
I also don't think you can actually modify the schema with it- just its data
cc-ing annavara@adobe.com

> To learn more about API Mesh, follow [this walkthrough](https://developer.adobe.com/graphql-mesh-gateway/gateway/mesh_walkthrough/) where you create, query, and transform data using a sample mesh.

Reviewer 2: I think this statement is a bit off
you can say 'use this sample API Mesh Config to.. create, query, and transform data (a mesh is not an industry standard word BTW)
Note that's not clear why would I use API Mesh to create API and query
the transform data (by not touching the vendor API - now that's interesting)
the create API and query - that's interesting from a single endpoint POV - usually in a graphql supergraph setting - because GraphQL language is selective like SQL - 
so you can combine multiple vendors API - create a supergraph with it  - so the consumer (the frontend) can interact with it and let API mesh deal with the underlying complexity - or at least separate concerns and have separate teams handle the wiring the API - usually without touching the vendor API (the multiple sources that API mesh uses to create the super graph - you usually can't modify or shouldn't the vendor API - you do that in this orchestration layer which is provided by API Mesh)

### Adobe I/O Events

[Adobe I/O Events for Adobe Commerce](https://developer.adobe.com/commerce/events/get-started/) allows you to create event-driven applications that take action when a shopper performs an action on a Commerce storefront, such as adding a product to a cart, clicking the Buy button, or creating an account. Commerce can now make transactional data available to applications created with Adobe App Builder.

In addition, [Conditional Events](https://developer.adobe.com/commerce/events/get-started/conditional-events/) give you the power to define rules that execute workflows when certain thresholds are met. For example, you could create a rule that sends a special coupon code for any new customer from a specified state.

## Combining extensibility components

> While App Builder, API Mesh, and Adobe I/O Events provide powerful extensibility on their own, combining them allows Adobe Commerce users to maximize their extensibility. The next few sections describe possible example implementations of this type of integration.

Reviewer 2: One way I like I like to look at it is from a consumer perspective.
The experience (the storefront) or integration consumer (ERP, CRM) needs to consume some new data.
- this data in order to be available back to the experience, it will need to be computed a certain business logic usually implemented in some backend system, not the experience or consumer side, it will have to compute and/or modify the behavior of the system or vendor API the exposed the data, and interact back to the experience/consumer with that computation.

---

# Extensibility examples

There are many ways to extend [Adobe Commerce](../index.md) today, such as [APIs](https://developer.adobe.com/commerce/webapi/) and [extensions](https://developer.adobe.com/commerce/marketplace/guides/sellers/extension-create/). However, these options in isolation only encapsulate a small fraction of possibilities with extensibility. This article speculates on what is possible with new and emergent technologies and more elaborate integrations to explore what can be created in the future.

## Understanding extensibility components

Adobe has developed several new technologies that can further extend the Adobe Commerce platform. Each of the following services has unique features and capabilities, which are important to understand individually before considering more complex examples.

### Adobe Developer App Builder

[App Builder](https://developer.adobe.com/app-builder/docs/overview/) is a complete framework that allows you to build and deploy custom web applications to extend Adobe Commerce and other Adobe solutions while running on Adobe infrastructure.

> To learn more about App Builder, follow [this step-by-step tutorial](https://developer.adobe.com/app-builder/docs/getting_started/first_app/) on creating your first app. These instructions will help you build foundational knowledge of App Builder and the tools it provides. You will learn how to create an app using the Adobe I/O CLI, and how to organize your code and projects within the developer console.

Reviewer: The first sentence says "follow" but the rest of the paragraph is future tense.

### API Mesh for Adobe Developer App Builder

> [API Mesh](https://developer.adobe.com/graphql-mesh-gateway/) is an Adobe service that allows you to take multiple vendor APIs of different protocols, like GraphQL, REST APIs, JSON Schemas, and SOAP implementations, and combine them into a single unified GraphQL endpoint. By consuming a mesh's unified GraphQL endpoint, you can connect Adobe Commerce to App Builder, Adobe I/O Events for Adobe Commerce, and any other service that provides an accessible source. You can further customize your experience by using [transforms](https://developer.adobe.com/graphql-mesh-gateway/gateway/transforms/) to customize the source data to your needs.

Reviewer 3: ```suggestion
[API Mesh](https://developer.adobe.com/graphql-mesh-gateway/) is an Adobe service that allows you to take multiple vendor APIs of different protocols, like GraphQL, REST, JSON schemas, and SOAP implementations, and combine them into a single unified GraphQL endpoint. By consuming a mesh's unified GraphQL endpoint, you can connect Adobe Commerce to App Builder, Adobe I/O Events for Adobe Commerce, and any other service that provides an accessible source. You can further customize your experience by using [transforms](https://developer.adobe.com/graphql-mesh-gateway/gateway/transforms/) to customize the source data to your needs.
```

![API Mesh diagram](../_images/api-mesh-diagram.png)
<!-- [Link to .mmd file](../_images/api-mesh-diagram.mmd) -->

API Mesh adds extensibility at the web API layer (headless/consumer level) by providing the relevant data and behavior from different vendor APIs to the consumer (usually an experience like a storefront, web store, or kiosk) without needing to modify these sources.

Additionally, the unified GraphQL endpoint allows you to integrate with other third-party systems (like ERPs and CRMs) or synchronize data between systems.

API Mesh has several features which allow for further extensibility:

- `beforeAll` [Hooks](https://developer.adobe.com/graphql-mesh-gateway/gateway/hooks/) - Allow you to insert a function before the query takes place, which can be helpful for use cases like authenticating or checking for a header before making a request.

> - [Custom Resolvers](https://developer.adobe.com/graphql-mesh-gateway/gateway/extending-unified-schema/) - Allow you to upload a resolver to your mesh that uses JavaScript to modify data, insert custom logic, and return custom computed data. In [this example](https://developer.adobe.com/graphql-mesh-gateway/gateway/extending-unified-schema/#programmatic-additionalresolvers), we show how you could use a discounts API to apply discounts to your products.

Reviewer: ```suggestion
- [Custom Resolvers](https://developer.adobe.com/graphql-mesh-gateway/gateway/extending-unified-schema/) - Allow you to upload a resolver to your mesh that uses JavaScript to modify data, insert custom logic, and return custom computed data. [This example](https://developer.adobe.com/graphql-mesh-gateway/gateway/extending-unified-schema/#programmatic-additionalresolvers) uses a discounts API to apply discounts to your products.
```

> - [Custom Resolvers](https://developer.adobe.com/graphql-mesh-gateway/gateway/extending-unified-schema/) - Allow you to upload a resolver to your mesh that uses Javascript to modify data, insert custom logic, and return custom computed data. In [this example](https://developer.adobe.com/graphql-mesh-gateway/gateway/extending-unified-schema/#programmatic-additionalresolvers), we show how you could use a discounts API to apply discounts to your products.

Reviewer 3: ```suggestion
- [Custom Resolvers](https://developer.adobe.com/graphql-mesh-gateway/gateway/extending-unified-schema/) - Allow you to upload a resolver to your mesh that uses JavaScript to modify data, insert custom logic, and return custom computed data. In [this example](https://developer.adobe.com/graphql-mesh-gateway/gateway/extending-unified-schema/#programmatic-additionalresolvers), we show how you could use a discounts API to apply discounts to your products.
```

> - [Transforms](https://developer.adobe.com/graphql-mesh-gateway/gateway/transforms/) in API Mesh allow you to modify the structure of an API's data without modifying the source itself.

Reviewer 3: ```suggestion
- [Transforms](https://developer.adobe.com/graphql-mesh-gateway/gateway/transforms/) - Allow you to modify the structure of an API's data without modifying the source itself.
```

To learn more about API Mesh, follow [this walkthrough](https://developer.adobe.com/graphql-mesh-gateway/gateway/mesh_walkthrough/) where you create, query, and transform data using a sample configuration.

### Adobe I/O Events

[Adobe I/O Events for Adobe Commerce](../events/index.md) allows you to react to an event using a runtime function when a shopper performs an action on a Commerce storefront, such as adding a product to a cart, clicking the Buy button, or creating an account. Commerce can now make transactional data available to applications created with Adobe App Builder.

In addition, [Conditional Events](../events/conditional-events.md) give you the power to define rules that execute workflows when certain thresholds are met. For example, you could create a rule that sends a special coupon code for any new customer from a specified state.

## Combining extensibility components

While App Builder, API Mesh, and Adobe I/O Events provide powerful extensibility on their own, combining them allows Adobe Commerce users to maximize their extensibility. The next few sections describe possible example implementations of this type of integration.

> The Commerce storefront acts as a consumer, while API Mesh acts as an orchestration layer. The consumer sends events to the orchestration layer, which can access backend systems and vendor APIs, then compute or modify the data and return it to the storefront or pass it on to the single-page application (App Builder). The consumer or storefront can also contact the single-page application directly.

Reviewer: This sentence is long and has a comma problem. I see that the subject (consumer) sends, computes, and returns, but it's very confusing to read with the which, ands, and ors. Maybe the first time you mention orchestration layer you can say which... not sure.

> The Commerce storefront acts as a consumer, while API Mesh acts as an orchestration layer. The consumer sends events to the orchestration layer which can access backend systems and vendor APIs then compute or modify the data and return it to the storefront or pass it on to the single-page application (App Builder). The consumer or storefront can also contact the single-page application directly.

Reviewer 3: ```suggestion
The Commerce storefront acts as a consumer, while API Mesh acts as an orchestration layer. The consumer sends events to the orchestration layer, which can access backend systems and vendor APIs, then compute or modify the data and return it to the storefront or pass it on to the single-page application (App Builder). The consumer or storefront can also contact the single-page application directly.
```

To learn more about App Builder, follow [this step-by-step tutorial](https://developer.adobe.com/app-builder/docs/getting_started/first_app/) on creating your first app. This provides you with the foundational knowledge of what App Builder is and how it works.

### API Mesh for Adobe Developer App Builder

[API Mesh](https://developer.adobe.com/graphql-mesh-gateway/) is an Adobe service that allows you to take multiple disparate sources, such as GraphQL endpoints, REST APIs, JSON Schemas, and SOAP implementations, and combine them into a single queryable GraphQL endpoint. By creating a mesh, you can connect Adobe Commerce to App Builder, Adobe IO Gateway, and any other service that provides an accessible source. You can further customize your experience by [transforming](https://developer.adobe.com/graphql-mesh-gateway/gateway/transforms/) source data.

> ![API Mesh diagram](../_images/api-mesh-diagram.png)

Reviewer 3: Very cool @jhadobe. I know we can't use mermaid.js natively, so adding the commented code for the diagram here while embedding a PNG format is clever.

Author: unfortunately when using the `<--!` commenting format, a right facing arrow in Mermaid `-->` will break the comment. So if we use these in the future, it might make more sense to just drop a <image>.js in the `/images` folder.

Reviewer: You can store them as MMD files in the images folder. And then with the right plugin, you can preview them in VScode.

Reviewer 4: I think your graphic should include REST at the expense of SOAP or API. Maybe API should be 3rd-party API?

Author: i went through the image with Cristian and updated the labels
